src/validation.cpp

```diff
@@ -3189,7 +3189,7 @@ std::vector<unsigned char> GenerateCoinbaseCommitment(CBlock& block, const CBloc
     std::vector<unsigned char> commitment;
     int commitpos = GetWitnessCommitmentIndex(block);
     std::vector<unsigned char> ret(32, 0x00);
-    if (consensusParams.vDeployments[Consensus::DEPLOYMENT_SEGWIT].nTimeout != 0) {
+    if (IsWitnessEnabled(pindexPrev, consensusParams) && !block.vtx[0]->HasWitness()) {
```

Comment:
It appears to me that here we do the output first in GenerateCoinbaseCommitment() then do the input next in UpdateUncommittedBlockStructures(), so checking HasWitness() here does not seem to make sense.